src/data_distribution/first_touch_distribution.hpp

```diff
@@ -0,0 +1,131 @@
+				const std::size_t rank, const std::size_t disp);
+			void _compare_exchange_dir(const void* desired, const void* expected, void* output_buffer,
+				const std::size_t size, const std::size_t rank, const std::size_t disp);
+		} // namespace atomic
+	} // namespace backend
+} // namespace argo
+
+#include "../backend/backend.hpp"
+
+/** @brief protects the owner directory */
+static pthread_mutex_t ownermutex = PTHREAD_MUTEX_INITIALIZER;
+
+namespace argo {
+	namespace data_distribution {
+		/**
+		 * @brief the first-touch data distribution
+		 * @details gives ownership of a page to the node that first touched it
+		 */
+		template<int instance>
+		class first_touch_distribution : public base_distribution<instance> {
+			private:
+				/**
+				 * @brief try and claim ownership of a page
+				 * @param addr address in the global address space
+				 * @return the homenode of addr
+				 */
+				static std::size_t first_touch (const std::size_t& addr);
+			public:
+				virtual node_id_t homenode (char* const ptr) {
+					std::size_t homenode;
+					const std::size_t addr = ptr - base_distribution<instance>::start_address;
+					const std::size_t index = 2 * (addr / granularity);
+					pthread_mutex_lock(&ownermutex);
+					argo::backend::atomic::_load_local_dir(&homenode, argo::backend::node_id(), index);
+					if (!homenode) { homenode = first_touch(addr); }
+					pthread_mutex_unlock(&ownermutex);
+
+					int n;
+					for(n = 0; n < base_distribution<instance>::nodes; n++) {
+						if((1UL << n) == homenode) {
+							homenode = n;
+						}
+					}
+
+					if(homenode >=(std::size_t)base_distribution<instance>::nodes) {
+						exit(EXIT_FAILURE);
+					}
+					return homenode;
+				}
+
+				virtual std::size_t local_offset (char* const ptr) {
+					std::size_t offset;
+					const std::size_t addr = ptr - base_distribution<instance>::start_address;
+					const std::size_t drift = addr % granularity;
+					const std::size_t index = 2 * (addr / granularity);
+					pthread_mutex_lock(&ownermutex);
+					argo::backend::atomic::_load_local_dir(&offset, argo::backend::node_id(), index+1);
+					offset += drift;
+					pthread_mutex_unlock(&ownermutex);
+
+					if(offset >=(std::size_t)base_distribution<instance>::size_per_node) {
+						exit(EXIT_FAILURE);
+					}
+					return offset;
+				}
+		};
+
+		template<int instance>
+		std::size_t first_touch_distribution<instance>::first_touch (const std::size_t& addr) {
+			/* variables for CAS */
+			node_id_t homenode;
+			std::size_t result;
+			constexpr std::size_t compare = 0;
+			const std::size_t id = 1 << argo::backend::node_id();
+			const std::size_t index = 2 * (addr / granularity);
+            
+			/* check/try to acquire ownership of the page with CAS to process' 0 index */
+			argo::backend::atomic::_compare_exchange_dir(&id, &compare, &result, sizeof(long), 0, index);
+
+			/* this process was the first one to deposit the id */
+			if (result == 0) {
+				homenode = id;
```

Comment:
It may be worth being careful here (in general for all distributions) to avoid confusion, `homenode` in most places of the backend refers to the ID returned by `argo::node_id()` (and equivalently `getHomenode(...)`, `getID()` and `workrank` in the MPI backend) which is in integer format (0 to n-1 for n nodes), while `id` here is the "bit representation". I suggest using a specific name for the bit representation such as `homenode_id` or maybe even `homenode_bit_id` when referring to the bit representation, and `homenode` when referring to the plain integer representation.

Reply:
Changed as requested.
We distinguish `homenode` from `homenode_bit_id`.